onnxruntime/core/framework/tensorprotoutils.cc

```diff
@@ -4,213 +4,577 @@
+    for (size_t i = 0, n = tensor_size; i < n; ++i) {
+      new (ptr + i) std::string();
+    }
+  } catch (std::exception& ex) {
+    return OrtCreateStatus(ORT_RUNTIME_EXCEPTION, ex.what());
+  }
+  return nullptr;
+}
+
+ORT_API(void, OrtUninitializeBuffer, _In_opt_ void* input, size_t input_len, enum ONNXTensorElementDataType type) {
+  if (type != ONNX_TENSOR_ELEMENT_DATA_TYPE_STRING || input == nullptr) return;
+  size_t tensor_size = input_len / sizeof(std::string);
+  std::string* ptr = reinterpret_cast<std::string*>(input);
+  using std::string;
+  for (size_t i = 0, n = tensor_size; i < n; ++i) {
+    ptr[i].~string();
+  }
+}
+
+static void DeleteHeapBuffer(void* param) noexcept {
+  UnInitializeParam* p = reinterpret_cast<UnInitializeParam*>(param);
+  OrtUninitializeBuffer(p->preallocated, p->preallocated_size, p->ele_type);
+  delete p;
+}
+
+Status TensorProtoToMLValue(const Env& env, const ORTCHAR_T* tensor_proto_path,
+                            const ONNX_NAMESPACE::TensorProto& tensor_proto, const MemBuffer& m, MLValue& value,
+                            OrtDeleter& deleter) {
+  const OrtAllocatorInfo& allocator = m.GetAllocInfo();
+  void* preallocated = m.GetBuffer();
+  size_t preallocated_size = m.GetLen();
+  ONNXTensorElementDataType ele_type = utils::GetTensorElementType(tensor_proto);
+  if (preallocated != nullptr && ele_type == ONNX_TENSOR_ELEMENT_DATA_TYPE_STRING) {
+    OrtStatus* status = OrtInitializeBufferForTensor(preallocated, preallocated_size, ele_type);
+    if (status != nullptr) {
+      OrtReleaseStatus(status);
+      return Status(common::ONNXRUNTIME, common::FAIL, "initialize preallocated buffer failed");
+    }
+    deleter.f = DeleteHeapBuffer;
+    deleter.param = new UnInitializeParam{preallocated, preallocated_size, ele_type};
+  } else {
+    deleter.f = nullptr;
+    deleter.param = nullptr;
+  }
   std::unique_ptr<Tensor> p_tensor;
-  ORT_RETURN_IF_ERROR(GetTensorFromTensorProto(input, &p_tensor, allocator, preallocated, preallocated_size));
+  std::string raw_data_from_file;
+  const void* raw_data = nullptr;
+  size_t raw_data_len = 0;
+  {
+    std::vector<int64_t> tensor_shape_vec = GetTensorShapeFromTensorProto(tensor_proto);
+    // Note: We permit an empty tensor_shape_vec, and treat it as a scalar (a tensor of size 1).
+    TensorShape tensor_shape{tensor_shape_vec};
+    if (tensor_proto.data_location() == TensorProto_DataLocation_EXTERNAL) {
+      if (ele_type == ONNX_TENSOR_ELEMENT_DATA_TYPE_STRING)
+        return Status(common::ONNXRUNTIME, common::FAIL, "string tensor can not have raw data");
+
+      std::unique_ptr<ExternalDataInfo> external_data_info;
+      ORT_RETURN_IF_ERROR(ExternalDataInfo::Create(tensor_proto.external_data(), &external_data_info));
+      if (external_data_info->GetOffset() > 0) {
+        return ORT_MAKE_STATUS(ONNXRUNTIME, NOT_IMPLEMENTED, "Cannot support tensor data with offset > 0");
+      }
```

Comment:
What's the reason why we don't support an offset > 0?

Reply:
Just not implemented.  It's rather complicated. And nobody would need it unless his file system has max file size limitation and he hit it.

---
In reply to: [261432252](https://github.com/Microsoft/onnxruntime/pull/520#discussion_r261432252) [](ancestors = 261432252)